pkg/distributor/streamsharder.go

```diff
@@ -0,0 +1,45 @@
+package distributor
+
+import (
+	"sync"
+
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+type streamSharder struct {
+	mu      sync.Mutex
+	streams map[string]int
+}
+
+func NewStreamSharder() StreamSharder {
+	return &streamSharder{
+		streams: make(map[string]int),
+	}
+}
+
+func (s *streamSharder) ShardsFor(stream logproto.Stream) (int, bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	shards := s.streams[stream.Labels]
```

Comment:
Since you're just reading here, a `RWMutex` would work and maybe reduce some contention